## weather 설명
- weather0: 위치 NaN 제거, 경기도 지역 제외, [풍향1, 풍향1(km/h), 풍향10, 풍향10(km/h) NaN값 날려버리기]
- weather1: 일강수, 강수60의 '.'을 0으로 대체, 라벨링을 위해 object type을 float type으로 바꿈
- weather2: 풍속10(km/h)를 풍속10(m/s)로 바꿔버림
- weather3: 풍속1(m/s)을 시간당 평균 내서 풍속 60(m/s)으로 바꿈.

## 문제점
- 풍속1 또는 풍속10일 때는 최대값이 12.61로 나왔는데 풍속60으로 평균치를 내버리니깐 '바람많이'라는 라벨링이 사라져 버리고 최댓값이 4.2
- 이게 과연 좋은 건지 나쁜건지.. 생각해보쟈

-------------------

# 예측 (지도학습)
- target = shared
- feature = hit 수?(유동인구) / 우천(안비, 적은비, 많은비) / 풍속 / 미세먼지 / 각 정거장 별 대여한 고객의 평균 목적지수 / 시간대 (기준 세울 필요. 일단 1시간) / 버스정거장 연계성 지수
- feature 정제해서 만들기
- 예측에 쓰이는 학습 model 공부하기

### 우천 라벨링?
- 맑음 : 0
- 적은비(~시간당 5mm) : 1
- 많은비(>시간당 5mm ) : 2

### 불쾌지수 라벨링? (X)
- [불쾌지수=1.8×기온–0.55×(1–습도)×(1.8×기온–26)+32]
- 불쾌지수 < 70 : 0
- 70 <= 불쾌지수 < 80: 1
- 불쾌지수 > 80 : 2

### 온도 라벨링?
- 냅두고 스케일링..?

### 풍속 라벨링?
- (풍속10(m/s) 기준) & (아래에 설명 더 잘 해놓음)
- 바람조금(0.0~1.5) 0
- 바람적당히(1.6~7.9) 1
- 바람많이(8.0~13.8) 2

### 미세먼지 라벨링?
- 호진오빠가 함

-------

In [1]:
import pandas as pd
import os
import dask.dataframe as dd
import progressbar

In [2]:
bar = progressbar.ProgressBar()

path_dir = 'weather/'
weatherlist = os.listdir(path_dir)

weather = pd.DataFrame()
c = []
for i in bar(weatherlist):
    k = dd.read_csv('weather/'+ i, engine = 'python', encoding = 'utf-8')
    c.append(k)
    
weather = dd.concat(c)

100% |########################################################################|


In [3]:
weather = weather.compute(scheduler = 'processes', num_workers = 4)

In [4]:
preprocess = pd.read_csv('preprocessed_final.csv')

In [5]:
from datetime import datetime
preprocess['time'] = pd.to_datetime(preprocess.time)
preprocess['time2'] = preprocess.time.dt.strftime('%H:%M')

weather['time'] = pd.to_datetime(weather.time)
weather['time2'] = weather.time.dt.strftime('%H:%M')

In [6]:
preprocess.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32956774 entries, 0 to 32956773
Data columns (total 10 columns):
Unnamed: 0           int64
parkingBikeTotCnt    float64
rackTotCnt           int64
shared               float64
stationId            object
stationLatitude      float64
stationLongitude     float64
stationName          object
time                 datetime64[ns]
time2                object
dtypes: datetime64[ns](1), float64(4), int64(2), object(3)
memory usage: 2.5+ GB


In [7]:
weather.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 751374 entries, 0 to 33
Data columns (total 23 columns):
Unnamed: 0    751374 non-null int64
ID            751374 non-null int64
지점            751374 non-null object
고도            751374 non-null object
강수            751374 non-null object
강수15          751374 non-null object
강수60          751374 non-null object
강수3H          751374 non-null object
강수6H          751374 non-null object
강수12H         751374 non-null object
일강수           751374 non-null object
기온            751374 non-null object
풍향1           733410 non-null float64
풍향1_          733410 non-null object
풍속1(km/h)     732099 non-null float64
풍향10          727676 non-null float64
풍향10_         727676 non-null object
풍속10(km/h)    727724 non-null float64
습도            751373 non-null object
해면기압          751373 non-null object
위치            751373 non-null object
time          751374 non-null datetime64[ns]
time2         751374 non-null object
dtypes: datetime64[ns](1), float

In [8]:
weather[weather.습도.isna()]

Unnamed: 0   ID  지점   고도 강수 강수15 강수60 강수3H 강수6H 강수12H  ... 풍향1_ 풍속1(km/h)  \
11          12  416  은평  58m  ●    0  0.5  5.5   24  44.5  ...   SE       5.0   

     풍향10 풍향10_  풍속10(km/h)   습도 해면기압   위치                time  time2  
11  139.0    SE        12.2  NaN  NaN  NaN 2019-08-12 11:16:15  11:16  

[1 rows x 23 columns]

-------------------

## weather0

In [49]:
weather0 = weather.copy()

In [50]:
weather0 = weather0[['ID', '지점', '강수', '강수15', '강수60', '강수3H', '강수12H', '일강수', '기온',
                     '풍향1', '풍속1(km/h)', '풍향10', '풍속10(km/h)', '위치', 'time', 'time2']]

In [45]:
weather0.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 751374 entries, 0 to 33
Data columns (total 16 columns):
ID            751374 non-null int64
지점            751374 non-null object
강수            751374 non-null object
강수15          751374 non-null object
강수60          751374 non-null object
강수3H          751374 non-null object
강수12H         751374 non-null object
일강수           751374 non-null object
기온            751374 non-null object
풍향1           733410 non-null float64
풍속1(km/h)     732099 non-null float64
풍향10          727676 non-null float64
풍속10(km/h)    727724 non-null float64
위치            751373 non-null object
time          751374 non-null datetime64[ns]
time2         751374 non-null object
dtypes: datetime64[ns](1), float64(4), int64(1), object(10)
memory usage: 97.5+ MB


#### 위치가 NaN인 값을 서울특별시 은평구 광진구로 바꿈. (이유가 다 있음)

In [73]:
weather0.reset_index(inplace=True)

In [21]:
weather0[weather0.ID == 416]

ID  지점 강수 강수15 강수60 강수3H 강수12H   일강수    기온    풍향1  풍속1(km/h)   풍향10  \
11  416  은평  ○    0    0    0     0     0    35   39.0        5.8   39.4   
11  416  은평  ○    0    0    0     0     0  27.6  176.6       10.4  128.2   
11  416  은평  ○    0    0    0  17.5     0  24.4    0.0        0.0    0.0   
11  416  은평  ○    0    0    0   2.5     0  24.4  251.3        2.2  184.7   
11  416  은평  ○    0    0    0     0     0    32   87.1        2.2   70.5   
11  416  은평  ○    0    0    0   6.0  46.5  28.0  120.1        4.0  131.0   
11  416  은평  ○    0    0    0     0     0  26.9    0.0        0.4    0.0   
11  416  은평  ○    0    0    0     0     0  25.6  320.3        3.2   63.3   
11  416  은평  ○    0    0    0     0     0  29.2  149.3        2.2  151.8   
11  416  은평  ○    0    0    0     0     0  23.5  151.9        4.7  168.9   
11  416  은평  ○    0    0    0     0     0  30.8   81.7       10.1   80.4   
11  416  은평  ●    0    1  1.5   1.5   1.5  22.6  300.0        1.1  332.0   
11  416  은평  ○    0    0    0     0     0  31.0  110.4        3.6  100.2   
11  416  은평  .    .    .    .     .     .     .    NaN        NaN    NaN   
11  416  은평  ○    0    0    0     0     0  23.1    0.0        0.4    0.0   
11  416  은평  ○    0    0    0     0     0  31.1   15.9        5.8   43.7   
11  416  은평  ○    0    0    0   8.0   8.0  24.7  188.3        1.8  158.6   
11  416  은평  ○    0    0    0     0     0  26.7  134.8        7.6  132.8   
11  416  은평  ○    0    0    0     0     0  34.3   65.0       11.9  103.3   
11  416  은평  ○    0    0    0     0     0  28.4  288.7        1.4  126.1   
11  416  은평  ○    0    0    0   2.5   2.5  24.7    0.0        0.4  146.3   
11  416  은평  ○    0    0    0  15.0     0  24.7   57.9        3.2   66.0   
11  416  은평  ○    0    0    0     0     0  31.7  133.0       14.4  113.3   
11  416  은평  ○    0    0    0     0     0  27.8   94.8        9.0  109.3   
11  416  은평  ○    0    0    0     0     0  28.4  222.3       12.2  256.6   
11  416  은평  ○    0    0    0  15.0     0  27.9  237.9        3.6  135.0   
11  416  은평  ○    0    0    0     0     0    33   64.0        4.3   69.7   
11  416  은평  ○    0    0    0  13.0  13.0  24.2    0.0        0.0  104.4   
11  416  은평  ○    0    0    0     0     0  30.5   45.8        2.2   64.8   
11  416  은평  ○    0    0    0     0     0  20.9  168.4        1.8  139.6   
..  ...  .. ..  ...  ...  ...   ...   ...   ...    ...        ...    ...   
11  416  은평  ○    0    0    0     0     0  27.3  153.1       14.4  120.4   
11  416  은평  ○    0    0    0  15.0     0  24.4  307.0        2.5  327.0   
11  416  은평  ○    0    0    0     0     0    28  238.7        6.1  255.0   
11  416  은평  ○    0    0    0  13.0     0  23.7    0.0        0.0  120.6   
11  416  은평  ○    0    0    0   6.0  46.5  28.8  157.5        8.6  141.9   
11  416  은평  ○    0    0    0   2.5     0  23.7    0.0        0.0    0.0   
11  416  은평  ○    0    0    0     0     0  30.6   72.3        7.9   54.1   
11  416  은평  ○    0    0    0   8.0   8.0  27.8    0.0        0.4  356.8   
11  416  은평  ○    0    0    0     0     0  23.9  141.6        3.2  155.3   
11  416  은평  ○    0    0    0   5.0     0  22.2   57.0        1.4   59.2   
11  416  은평  ○    0    0    0     0     0  34.4   83.9        7.2   81.8   
11  416  은평  ○    0    0    0     0     0  28.1  155.4        1.1  121.3   
11  416  은평  ○    0  0.5  0.5   0.5   0.5  25.2  160.0        1.8  166.4   
11  416  은평  ○    0    0    0     0     0  30.8   56.7        4.3   63.3   
11  416  은평  ○    0    0    0     0     0  23.5  272.5        0.7  297.0   
11  416  은평  ○    0    0    0     0     0  18.1  272.4        2.2  201.7   
11  416  은평  ○    0    0    0     0     0  30.3   63.3        2.5   58.7   
11  416  은평  ○    0    0    4    29  46.5    29  172.1        6.8  156.2   
11  416  은평  ○    0    0    0     0     0  25.1    0.0        0.0  122.5   
11  416  은평  ○    0  0.5    3  24.5  17.5  26.2  102.4        5.0  101.7   
11  416  은평  ○    0    0    0     0     0  31.1   52.

In [52]:
weather0.위치 = weather0.위치.fillna('서울특별시 은평구 진관동')

In [74]:
weather0

index   ID     지점 강수 강수15 강수60 강수3H 강수12H 일강수    기온    풍향1  풍속1(km/h)  \
0           0  108     서울  ○    0    0    0     0   0  36.2   40.8        5.0   
1           1  419     중구  ○    0    0    0     0   0    34  139.3        6.5   
2           2  415   용산 *  ○    0    0    0     0   0  36.7   42.3       12.2   
3           3  421     성동  ○    0    0    0     0   0  34.9    8.6        8.3   
4           4  413     광진  ○    0    0    0     0   0  35.6   63.5        8.6   
5           5  408    동대문  ○    0    0    0     0   0  35.6   78.8       14.0   
6           6  409     중랑  ○    0    0    0     0   0  36.7   93.3       11.9   
7           7  414     성북  ○    0    0    0     0   0  35.6   33.3       10.8   
8           8  424     강북  ○    0    0    0     0   0  36.3   61.8        8.6   
9           9  406     도봉  ○    0    0    0     0   0    36  184.1        2.5   
10         10  407     노원  ○    0    0    0     0   0  35.2   52.5        6.5   
11         11  416     은평  ○    0    0    0     0   0    35   39.0        5.8   
12         12  412    서대문  ○    0    0    0     0   0  35.7  198.1        0.7   
13         13  411     마포  ○    0    0    0     0   0    36   71.7        7.6   
14         14  405   양천 *  ○    0    0    0     0   0  37.3  333.6        3.2   
15         15  404     강서  ○    0    0    0     0   0  35.7    7.4        9.4   
16         16  110  김포(공)  -    0    0    0     0   0  36.4   80.0        2.2   
17         17  423     구로  ○    0    0    0     0   0  36.7  351.3       11.9   
18         18  417     금천  ○    0    0    0     0   0  36.5   23.4        7.2   
19         19  418     한강  ○    0    0    0     0   0  34.9   86.5        7.6   
20         20  510  영등포 *  ○    0    0    0     0   0  36.7  310.9        5.4   
21         21  889    현충원  ○    0    0    0     0   0  37.2   27.4        5.8   
22         22  410    기상청  ○    0    0    0     0   0  36.7  112.3        6.8   
23         23  509     관악  ○    0    0    0     0   0  35.1    5.3        3.6   
24         24  425     남현  ○    0    0    0     0   0    35   32.1        9.7   
25         25  401     서초  ○    0    0    0     0   0  35.3  280.4        4.7   
26         26  400     강남  ○    0    0    0     0   0  37.1  110.9        6.5   
27         27  403   송파 *  ○    0    0    0     0   0  37.1   55.3        8.6   
28         28  402     강동  ○    0    0    0     0   0  34.3   63.7        1.8   
29          0  108     서울  ○    0    0    0     0   0  27.3   94.1        7.9   
...       ...  ...    ... ..  ...  ...  ...   ...  ..   ...    ...        ...   
641031     28  402     강동  ○    0    0    0     0   0  25.0  235.2        2.5   
641032      0  108     서울  ○    0    0    0     0   0    28   44.5        6.1   
641033      1  419     중구  ○    0    0    0     0   0  26.2   98.3       11.9   
641034      2  415   용산 *  ○    0    0    0     0   0  28.9   41.7       11.2   
641035      3  421     성동  ○    0    0    0     0   0  28.3   27.4        4.7   
641036      4  413     광진  ○    0    0    0     0   0    27  102.5        4.0   
641037      5  408    동대문  ○    0    0    0     0   0  27.6  104.4        4.7   
641038      6  409     중랑  ○    0    0    0     0   0    28   71.3        9.7   
641039      7  414     성북  ○    0    0    0     0   0  27.5   55.9        7.6   
641040      8  424     강북  ○    0    0    0     0   0  28.2   78.9        6.8   
641041      9  406     도봉  ○    0    0    0     0   0  26.5   41.8        1.8   
641042     10  407     노원  ○    0    0    0     0   0  26.6   68.7       13.0   
641043     11  416     은평  ○    0    0    0     0   0  27.8  166.3        7.2   
641044     12  412    서대문  ○    0    0    0     0   0  27.6  166.6        5.8   
641045     13  411     마포  ○    0    0    0     0   0  28.1   85.4       10.8   
641046     14  405   양천 *  ○    0    0    0     0   0  29.3   79.7        6.1   
641047     15  404     강서  ○    0    0    0     0   0  28.3   79.1        6.1   
641048     16  110  김포(공)  -    0    0 

In [56]:
weather0[weather0.위치.isna()]

Empty DataFrame
Columns: [ID, 지점, 강수, 강수15, 강수60, 강수3H, 강수12H, 일강수, 기온, 풍향1, 풍속1(km/h), 풍향10, 풍속10(km/h), 위치, time, time2]
Index: []

#### 경기도 위치를 다 빼버림

In [57]:
경기도 = ['광명', '소하', '학온동', '과천', '관악(레)']
for each in 경기도:
    weather0 = weather0[weather0.지점 != each]

In [58]:
weather0.지점.unique()

array(['서울', '중구', '용산 *', '성동', '광진', '동대문', '중랑', '성북', '강북', '도봉',
       '노원', '은평', '서대문', '마포', '양천 *', '강서', '김포(공)', '구로', '금천', '한강',
       '영등포 *', '현충원', '기상청', '관악', '남현', '서초', '강남', '송파 *', '강동'],
      dtype=object)

#### 풍향1, 풍향1(km/h), 풍향10, 풍향10(km/h)  NaN 날려버리기
- 64만 데이터 중에 3만 데이터만 사라지는 거니깐 괜찮아!

In [88]:
weather0 = weather0[weather0.풍향1.notnull()]
weather0 = weather0[weather0['풍속1(km/h)'].notnull()]
weather0 = weather0[weather0.풍향10.notnull()]
weather0 = weather0[weather0['풍속10(km/h)'].notnull()]

In [89]:
weather0.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 623198 entries, 0 to 641060
Data columns (total 17 columns):
index         623198 non-null int64
ID            623198 non-null int64
지점            623198 non-null object
강수            623198 non-null object
강수15          623198 non-null object
강수60          623198 non-null object
강수3H          623198 non-null object
강수12H         623198 non-null object
일강수           623198 non-null object
기온            623198 non-null object
풍향1           623198 non-null float64
풍속1(km/h)     623198 non-null float64
풍향10          623198 non-null float64
풍속10(km/h)    623198 non-null float64
위치            623198 non-null object
time          623198 non-null datetime64[ns]
time2         623198 non-null object
dtypes: datetime64[ns](1), float64(4), int64(2), object(10)
memory usage: 85.6+ MB


-------------------

## weather 1
### 우천 라벨링
- 맑음 : 0
- 적은비(~시간당 5mm) : 1
- 많은비(>시간당 5mm ) : 2
<br><br>
- 강수 ['○', '-', '●', '.'] -> 신경쓰지 말자
- 일강수랑 강수 60만 중요한 거 같음 (시간당 자료만 필요하니깐)
- 강수60, 일강수의 object type를 float로 변경

In [93]:
weather0.drop('index', axis=1, inplace=True)

In [115]:
weather1 = weather0.drop('강수', axis=1)

In [135]:
weather1['일강수'] = weather1['일강수'].replace('.', 0)
weather1['강수60'] = weather1['강수60'].replace('.', 0)

In [136]:
weather1[weather1.강수60== '.']

Empty DataFrame
Columns: [ID, 지점, 강수15, 강수60, 강수3H, 강수12H, 일강수, 기온, 풍향1, 풍속1(km/h), 풍향10, 풍속10(km/h), 위치, time, time2]
Index: []

In [140]:
weather1["강수60"] = weather1.강수60.astype(float)
weather1["일강수"] = weather1.일강수.astype(float)

In [141]:
weather1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 623198 entries, 0 to 641060
Data columns (total 15 columns):
ID            623198 non-null int64
지점            623198 non-null object
강수15          623198 non-null object
강수60          623198 non-null float64
강수3H          623198 non-null object
강수12H         623198 non-null object
일강수           623198 non-null float64
기온            623198 non-null object
풍향1           623198 non-null float64
풍속1(km/h)     623198 non-null float64
풍향10          623198 non-null float64
풍속10(km/h)    623198 non-null float64
위치            623198 non-null object
time          623198 non-null datetime64[ns]
time2         623198 non-null object
dtypes: datetime64[ns](1), float64(6), int64(1), object(7)
memory usage: 76.1+ MB


In [142]:
weather1['우천라벨링'] = weather1.강수60.apply(lambda x: 0 if x == 0 else
                                        1 if x <= 5 else 2)

#대여이력['weekend'] = 대여이력.대여일시.apply(lambda x: 0 if x.weekday() < 5 else 1)

In [144]:
weather1[weather1.강수60 > 5]

ID     지점  강수15  강수60  강수3H 강수12H   일강수    기온    풍향1  풍속1(km/h)  \
1140    406     도봉   1.5   6.5    23    64  58.0    25   18.0        0.7   
1284    424     강북     0   8.0  48.5  48.5  48.5  26.8   42.8        2.5   
1285    406     도봉     0  15.0  42.5  42.5  42.5     .  328.7        2.9   
1684    415   용산 *   2.0   7.0   9.0   9.0   9.0  24.8   54.9        6.5   
1685    421     성동   6.0  13.0  16.0  16.0  16.0  23.8   10.7       12.6   
1686    413     광진   1.5  15.5  26.0  26.0  26.0  23.8   26.9        5.8   
1687    408    동대문   5.0  16.5  19.0  19.0  19.0  24.1   40.5        9.4   
1688    409     중랑   3.5  17.5  22.5  22.5  22.5  24.4  349.1       13.0   
1691    406     도봉   1.5   5.5   7.0   7.0   7.0  23.7   12.1        7.2   
1692    407     노원   5.0  22.5  30.0  30.0  30.0  23.6  321.4        0.4   
1703    889    현충원   3.0  10.0  12.5  13.0  13.0  24.7  108.2        3.2   
1705    509     관악   6.0  10.5  12.0  12.0  12.0  23.2  352.8        4.7   
1707    401     서초   4.0  18.0  22.0  22.0  22.0  24.0  303.7        4.0   
1708    400     강남   1.0  13.5  22.5  23.0  23.0  24.4  336.9        0.7   
1709    403   송파 *   0.5  12.5  22.5  23.0  23.0  24.2   35.3        2.5   
1710    402     강동   1.0  11.5  23.5  25.0  25.0  23.9   17.2        7.6   
2792    424     강북     2   5.5     6    28  22.5  25.2  284.0        1.8   
2793    406     도봉     3   6.5   7.5  32.5  26.5  24.3  240.5        4.7   
2795    416     은평   4.5   9.0  10.5    34  27.0  24.2  170.3        4.3   
2796    412    서대문     3   7.0   7.5    25  19.5  24.5  201.5        2.2   
2797    411     마포   3.5   7.0     8  26.5  20.0    25    0.0        0.0   
2798    405   양천 *     3   9.0   9.5    25  18.0  25.2  123.6        2.5   
2799    404     강서     2   9.0  10.5  35.5  27.5    25  206.1        4.3   
2800    110  김포(공)   0.7   7.0   8.3  29.6  24.7  24.8  250.0        8.6   
2801    423     구로     2   8.0     9    28  18.5  24.9  189.3        5.0   
2802    417     금천   3.5   8.0   8.5  22.5  14.5  24.9  180.9        1.8   
2803    418     한강     2   6.0     6    22  15.5  25.5    0.0        0.0   
2804    510  영등포 *     3   6.5     7  22.5  16.0  25.2    0.0        0.4   
2806    410    기상청     3   7.0   7.5    23  14.5  24.8  250.6        5.4   
2807    509     관악     3   6.0     7    23  13.5    24  163.3        4.3   
...     ...    ...   ...   ...   ...   ...   ...   ...    ...        ...   
637162  415   용산 *     0   6.0   7.0  11.5  39.5  26.9   39.6        5.8   
637167  414     성북     0   8.0   8.0  17.5  64.5  25.9   14.8        0.7   
637171  416     은평     0  14.5  14.5  19.5  61.0  25.5   71.5        2.5   
637172  412    서대문     0  13.0  13.0  16.5  55.0    26   79.5        2.2   
637173  411     마포   0.5   7.5   7.5  12.5  52.5  26.6   97.3        8.3   
638121  413     광진     2   6.0     8    32  38.5  26.5   69.6        6.5   
638143  400     강남   2.5   9.0    10    34  42.0  27.3   79.6        3.6   
638144  403   송파 *     4  10.0    11    34  41.5  26.8   40.6        4.7   
638145  402     강동     7  10.5    11    33  39.5  26.4   14.1        6.5   
638840  403   송파 *     0   7.5   7.5   7.5   7.5  31.6  238.9        2.5   
639350  108     서울     4   5.6   5.6   5.6   5.6  29.0  100.2        7.9   
639351  419     중구    13  33.0  33.0  33.0  33.0  25.4  311.1        5.0   
639353  421     성동     4   6.5   6.5   6.5   6.5  26.9   41.9       12.2   
639354  413     광진    12  24.0  24.0  24.0  24.0  25.8  312.4        9.0   
639355  408    동대문   6.5  22.5  22.5  22.5  22.5  25.7  192.2       23.4   
639356  409     중랑  10.5  22.5  22.5  22.5  22.5  26.2  205.5       24.5   
639369  418     한강   3.5  22.0  22.0  22.0  22.0  27.4  114.0        4.3   
639372  410    기상청     0   8.5     .   8.5   8.5  27.5  142.2        1.8   
639378  402     강동   7.5   7.5   7.5   7.5   7.5  26.5  299.7        9.0   
639912  416     은평   9.5   9.5   9.5   9.5   9.5  25.9   81.5       11.2   
639923  410    기상청     0   6.0     7     7   7.0  32.4   

-------------------

## weather 2
### 풍속 라벨링
- km/h를 m/s로 변환
<br><br>

계급|명칭|지표면의 특징 및 풍속(m/sec)

0|고요|연기가 곧게 위로 올라감 (0.0~0.2)<br>
1|실바람|연기가 풀려서 오르고 해면은 물고기 비늘 모양의 잔 물결이 일어남 (0.3~1.5)<br>
2|남실바람|바람이 얼굴에 느껴지고 나뭇잎이 흔들리며 풍향계가 움직임 (1.6~3.3)<br>
3|산들바람|나뭇잎과 잔가지가 일정하게 흔들리고 깃발이 가볍게 휘날림 (3.4~5.4)<br>
4|건들바람|먼지가 일고 종잇조각이 날리며 나무의 잔가지가 움직임 (5.5~7.9)<br>
5|흔들바람|작은 나무 전체가 흔들리며 작은 물결이 호수에 생김 (8.0~10.7)<br>
6|된바람|큰 나뭇가지가 흔들리고 우산 쓰기가 힘들며, 큰 물결이 일기 시작함 (10.8~13.8)<br>
7|센바람|나무 전체가 흔들리고 걷기가 힘들며 해상의 물마루가 부서짐 (13.9~17.1)<br>
8|큰바람|잔가지가 꺾이고 걷기 힘들며 해상의 풍량이 높아지고 물보라가 읾 (17.2~20.7)<br>
9|큰센바람|건축물에 약간 피해가 있으며, 물보라가 소용돌이침 (20.8~24.4)<br>
10|노대바람|나무가 뽑히고 건축물에 상당한 피해가 있음 (24.5~28.4)<br>
11|왕바람|건축물에 큰 피해가 있음 (28.5~32.6)<br>
12|싹쓸바람|매우 큰 피해가 발생하며 배의 침몰이 염려됨 (32.7 이상)<br>

- 내 맘대로 정한 기준 **
- 바람조금(0.0~1.5) 0
- 바람적당히(1.6~7.9) 1
- 바람많이(8.0~13.8) 2

In [244]:
weather2 = weather1.copy()

In [245]:
weather2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 623198 entries, 0 to 641060
Data columns (total 16 columns):
ID            623198 non-null int64
지점            623198 non-null object
강수15          623198 non-null object
강수60          623198 non-null float64
강수3H          623198 non-null object
강수12H         623198 non-null object
일강수           623198 non-null float64
기온            623198 non-null object
풍향1           623198 non-null float64
풍속1(km/h)     623198 non-null float64
풍향10          623198 non-null float64
풍속10(km/h)    623198 non-null float64
위치            623198 non-null object
time          623198 non-null datetime64[ns]
time2         623198 non-null object
우천라벨링         623198 non-null int64
dtypes: datetime64[ns](1), float64(6), int64(2), object(7)
memory usage: 80.8+ MB


In [171]:
max(weather2['풍속10(km/h)'])

45.4

In [172]:
weather2.describe()

ID           강수60            일강수            풍향1  \
count  623198.000000  623198.000000  623198.000000  623198.000000   
mean      413.847224       0.287312       3.345996     154.956247   
std       122.581636       1.542788       9.666424     106.400965   
min       108.000000       0.000000       0.000000       0.000000   
25%       405.000000       0.000000       0.000000      66.000000   
50%       412.000000       0.000000       0.000000     140.400000   
75%       419.000000       0.000000       0.000000     239.100000   
max       889.000000      48.500000      72.500000     360.000000   

           풍속1(km/h)           풍향10     풍속10(km/h)          우천라벨링  
count  623198.000000  623198.000000  623198.000000  623198.000000  
mean        5.462423     157.738625       5.479646       0.107871  
std         4.493575     104.529735       4.043822       0.357565  
min         0.000000       0.000000       0.000000       0.000000  
25%         2.200000      70.800000       2.500000       0.000000  
50%         4.700000     142.800000       4.700000       0.000000  
75%         7.600000     238.900000       7.600000       0.000000  
max        50.800000     360.000000      45.400000       2.000000

#### 풍속10(km/h)를 풍속10(m/s)로 바꿔버림

In [173]:
def to_ms(n):
    '''input a km/hr number and convert it to m/s'''
    result = round((n * 1000)/3600, 2)
    return result

In [246]:
weather2['풍속1(km/h)'] = weather2['풍속1(km/h)'].apply(lambda x: to_ms(x))
weather2['풍속10(km/h)'] = weather2['풍속10(km/h)'].apply(lambda x: to_ms(x))

In [247]:
max(weather2['풍속10(km/h)'])

12.61

In [248]:
weather2 = weather2.rename(columns = {'풍속1(km/h)':'풍속1(m/s)'})
weather2 = weather2.rename(columns = {'풍속10(km/h)':'풍속10(m/s)'})

In [200]:
weather2.describe()

ID           강수60            일강수            풍향1  \
count  623198.000000  623198.000000  623198.000000  623198.000000   
mean      413.847224       0.287312       3.345996     154.956247   
std       122.581636       1.542788       9.666424     106.400965   
min       108.000000       0.000000       0.000000       0.000000   
25%       405.000000       0.000000       0.000000      66.000000   
50%       412.000000       0.000000       0.000000     140.400000   
75%       419.000000       0.000000       0.000000     239.100000   
max       889.000000      48.500000      72.500000     360.000000   

            풍속1(m/s)           풍향10      풍속10(m/s)          우천라벨링  
count  623198.000000  623198.000000  623198.000000  623198.000000  
mean        1.517316     157.738625       1.522099       0.107871  
std         1.248253     104.529735       1.123327       0.357565  
min         0.000000       0.000000       0.000000       0.000000  
25%         0.610000      70.800000       0.690000       0.000000  
50%         1.310000     142.800000       1.310000       0.000000  
75%         2.110000     238.900000       2.110000       0.000000  
max        14.110000     360.000000      12.610000       2.000000

In [257]:
weather2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 623198 entries, 0 to 641060
Data columns (total 16 columns):
ID           623198 non-null int64
지점           623198 non-null object
강수15         623198 non-null object
강수60         623198 non-null float64
강수3H         623198 non-null object
강수12H        623198 non-null object
일강수          623198 non-null float64
기온           623198 non-null object
풍향1          623198 non-null float64
풍속1(m/s)     623198 non-null float64
풍향10         623198 non-null float64
풍속10(m/s)    623198 non-null float64
위치           623198 non-null object
time         623198 non-null datetime64[ns]
time2        623198 non-null object
우천라벨링        623198 non-null int64
dtypes: datetime64[ns](1), float64(6), int64(2), object(7)
memory usage: 80.8+ MB


-------------------

## weather 3
#### 풍속 60(m/s)로 바꾸기 위해서..
- 한시간당(ex. 1:00~1:59 regardless of not having all the 59 minutes) 풍속1(m/s)의 평균을 구해서 풍속60(m/s)에 넣기

In [289]:
weather2.sort_values(by='time', ascending=False)

ID     지점 강수15  강수60 강수3H 강수12H  일강수    기온    풍향1  풍속1(m/s)   풍향10  \
295872  415   용산 *    0   0.0    0     0  0.0    31  221.3      2.81  218.6   
295879  406     도봉    0   0.0    0     0  0.0  30.1  342.6      1.00  341.6   
295873  421     성동    0   0.0    0     0  0.0  30.6  212.6      2.31  199.1   
295874  413     광진    0   0.0    0     0  0.0  31.6   87.2      1.81  151.2   
295875  408    동대문    0   0.0    0     0  0.0  29.7  156.6      3.39  152.3   
295876  409     중랑    0   0.0    0     0  0.0    31   67.2      2.11  124.6   
295877  414     성북    0   0.0    0     0  0.0  29.8   70.1      1.69   61.3   
295878  424     강북    0   0.0    0     0  0.0  30.4  145.9      2.00  102.0   
295880  407     노원    0   0.0    0     0  0.0  29.5   59.2      1.81   72.1   
295870  108     서울    0   0.0    0     0  0.0  30.8    2.5      1.19  335.0   
295881  416     은평    0   0.0    0     0  0.0  29.5   10.7      1.69   53.6   
295882  412    서대문    0   0.0    0     0  0.0  30.9   56.4      0.39   24.6   
295883  411     마포    0   0.0    0     0  0.0  30.2  291.0      2.61  304.8   
295884  405   양천 *    0   0.0    0     0  0.0  31.6   81.9      0.81    1.4   
295885  404     강서    0   0.0    0     0  0.0  30.3  355.9      2.11  346.6   
295886  110  김포(공)    .   0.0    .     .  0.0  30.4  200.0      1.50   60.0   
295871  419     중구    0   0.0    0     0  0.0  28.6  109.3      0.19   76.3   
295887  423     구로    0   0.0    0     0  0.0  31.2    5.7      1.00  324.8   
295895  401     서초    0   0.0    0     0  0.0  30.6  102.1      1.19  141.4   
295888  417     금천    0   0.0    0     0  0.0  30.7  154.8      1.81  195.5   
295889  418     한강    0   0.0    0     0  0.0  31.4    0.0      0.00    0.0   
295890  510  영등포 *    0   0.0    0     0  0.0  31.2  189.0      0.50  323.4   
295898  402     강동    0   0.0    0     0  0.0  29.3  102.1      0.61   21.1   
295897  403   송파 *    0   0.0    0     0  0.0  30.9  317.3      1.19  347.5   
295891  889    현충원    0   0.0    0     0  0.0  31.6   41.8      0.81   30.7   
295896  400     강남    0   0.0    0     0  0.0  32.9  180.7      0.69  347.6   
295892  410    기상청    0   0.0    0     0  0.0  31.9   47.7      0.81   50.0   
295893  509     관악    0   0.0    0     0  0.0    29  196.1      1.31  184.1   
295894  425     남현    0   0.0    0     0  0.0  30.4   91.4      1.81  112.7   
592877  413     광진    0   0.0    0     0  0.0  31.6   87.2      1.81  151.2   
...     ...    ...  ...   ...  ...   ...  ...   ...    ...       ...    ...   
175961  415   용산 *    0   0.0    0     0  0.0    30   48.6      1.50   27.5   
175962  421     성동    0   0.0    0     0  0.0    30  353.3      1.00   56.4   
175984  401     서초    0   0.0    0     0  0.0  29.8  175.6      0.11   64.1   
175964  408    동대문    0   0.0    0     0  0.0  28.9  118.5      1.39   82.3   
175986  403   송파 *    0   0.0    0     0  0.0  28.7  276.3      0.31  349.1   
175963  413     광진    0   0.0    0     0  0.0  27.9   88.7      1.31   81.7   
175987  402     강동    0   0.0    0     0  0.0  28.6   82.0      2.61   36.7   
74882   110  김포(공)    0   0.0    .     .  0.0  24.8  320.0      1.00  330.0   
401898  110  김포(공)    0   0.0    .     .  0.0  24.8  320.0      1.00  330.0   
148237  110  김포(공)    0   0.0    .     .  0.0  24.8  320.0      1.00  330.0   
368648  110  김포(공)    .   0.0    .     .  0.0  24.6  330.0      0.31  300.0   
142223  110  김포(공)    .   0.0    .     .  0.0  24.6  330.0      0.31  300.0   
411236  110  김포(공)    .   0.0    .     .  0.0  24.6  330.0      0.31  300.0   
525089  110  김포(공)    0   0.0    .     .  0.0  24.6  290.0      0.89  300.0   
247890  110  김포(공)    0   0.0    .     .  0.0  24.6  290.0      0.89  300.0   
621064  110  김포(공)    0   0.0    .     .  0.0  24.6  290.0      0.89  300.0   
95008   110  김포(공)    0   0.0    .     .  0.0  24.6  290.0      1.11  310.0   
404914  110  김포(공)    0   0.0    .     .  0.0  24.6  290.0      1.11  310.0   
273393  110  김포(공)    0   0.0    .     .  0.0  24.7  320.0 

In [293]:
k = pd.date_range(start='20190803', end='20190820', freq='H')

In [294]:
k

DatetimeIndex(['2019-08-03 00:00:00', '2019-08-03 01:00:00',
               '2019-08-03 02:00:00', '2019-08-03 03:00:00',
               '2019-08-03 04:00:00', '2019-08-03 05:00:00',
               '2019-08-03 06:00:00', '2019-08-03 07:00:00',
               '2019-08-03 08:00:00', '2019-08-03 09:00:00',
               ...
               '2019-08-19 15:00:00', '2019-08-19 16:00:00',
               '2019-08-19 17:00:00', '2019-08-19 18:00:00',
               '2019-08-19 19:00:00', '2019-08-19 20:00:00',
               '2019-08-19 21:00:00', '2019-08-19 22:00:00',
               '2019-08-19 23:00:00', '2019-08-20 00:00:00'],
              dtype='datetime64[ns]', length=409, freq='H')

In [285]:
weather2['풍속60(m/s)'] = None

In [304]:
bar = progressbar.ProgressBar()
weather3 = pd.DataFrame()
for i in bar(range(len(k)-1)):
    H_weather = weather2[(weather2.time >= k[i])&(weather2.time < k[i+1])]
    H_weather['풍속60(m/s)'] = H_weather['풍속1(m/s)'].mean()
    weather3 = pd.concat([weather3, H_weather], ignore_index=True)

/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """
100% |########################################################################|


In [305]:
weather3

ID     지점 강수15  강수60 강수3H 강수12H  일강수    기온    풍향1  풍속1(m/s)   풍향10  \
0       110  김포(공)    0   0.0    .     .  0.0  24.7  280.0      0.69  270.0   
1       110  김포(공)    0   0.0    .     .  0.0  24.7  280.0      0.69  270.0   
2       110  김포(공)    0   0.0    .     .  0.0  24.8  300.0      1.39  300.0   
3       110  김포(공)    0   0.0    .     .  0.0  24.8  320.0      1.00  330.0   
4       110  김포(공)    0   0.0    .     .  0.0  24.8  290.0      1.31  290.0   
5       110  김포(공)    0   0.0    .     .  0.0  24.6  290.0      1.11  310.0   
6       110  김포(공)    .   0.0    .     .  0.0  24.6  330.0      0.31  300.0   
7       110  김포(공)    0   0.0    .     .  0.0  24.8  320.0      1.00  330.0   
8       110  김포(공)    0   0.0    .     .  0.0  24.6  290.0      0.89  300.0   
9       110  김포(공)    0   0.0    .     .  0.0  24.7  310.0      1.61  310.0   
10      110  김포(공)    0   0.0    .     .  0.0  24.7  320.0      1.61  310.0   
11      110  김포(공)    0   0.0    .     .  0.0  24.7  310.0      1.61  310.0   
12      110  김포(공)    .   0.0    .     .  0.0  24.6  330.0      0.31  300.0   
13      110  김포(공)    0   0.0    .     .  0.0  24.8  320.0      1.00  330.0   
14      110  김포(공)    0   0.0    .     .  0.0  24.6  290.0      1.11  310.0   
15      110  김포(공)    .   0.0    .     .  0.0  24.6  330.0      0.31  300.0   
16      110  김포(공)    0   0.0    .     .  0.0  24.7  290.0      1.19  290.0   
17      110  김포(공)    0   0.0    .     .  0.0  24.7  310.0      1.61  310.0   
18      110  김포(공)    0   0.0    .     .  0.0  24.6  290.0      0.89  300.0   
19      110  김포(공)    0   0.0    .     .  0.0  24.8  300.0      1.39  300.0   
20      400     강남    .   0.0    0     0  0.0  26.8  286.9      0.89  284.5   
21      110  김포(공)    0   0.0    .     .  0.0  24.7  310.0      1.61  310.0   
22      110  김포(공)    0   0.0    .     .  0.0  24.6  290.0      0.89  300.0   
23      108     서울    0   0.0    0     0  0.0    29  133.0      0.69   89.4   
24      419     중구    0   0.0    0     0  0.0  28.6   67.1      1.81   66.6   
25      415   용산 *    0   0.0    0     0  0.0  29.9   67.9      1.61   32.7   
26      421     성동    0   0.0    0     0  0.0  30.1  299.4      0.00  227.1   
27      413     광진    0   0.0    0     0  0.0    28  129.1      0.89  114.0   
28      408    동대문    0   0.0    0     0  0.0  29.2  141.7      1.39  134.2   
29      409     중랑    0   0.0    0     0  0.0  29.6   88.9      1.50   49.4   
...     ...    ...  ...   ...  ...   ...  ...   ...    ...       ...    ...   
623168  402     강동    0   0.0    0     0  0.0  29.3   33.7      1.50   16.6   
623169  108     서울    0   0.0    0     0  0.0  30.7  323.3      1.39  340.3   
623170  419     중구    0   0.0    0     0  0.0  28.9   99.2      0.81   85.2   
623171  415   용산 *    0   0.0    0     0  0.0  30.9  229.6      2.69  212.2   
623172  421     성동    0   0.0    0     0  0.0  30.7  202.1      2.00  199.3   
623173  413     광진    0   0.0    0     0  0.0  31.8  116.5      1.19  153.5   
623174  408    동대문    0   0.0    0     0  0.0  29.8  176.1      1.89  126.7   
623175  409     중랑    0   0.0    0     0  0.0  31.4  117.3      2.19  117.5   
623176  414     성북    0   0.0    0     0  0.0  29.7  102.4      1.81   43.6   
623177  424     강북    0   0.0    0     0  0.0  31.1  128.2      1.89   89.3   
623178  406     도봉    0   0.0    0     0  0.0  30.3  312.0      1.89  357.9   
623179  407     노원    0   0.0    0     0  0.0  29.7   82.0      1.61   70.3   
623180  416     은평    0   0.0    0     0  0.0  29.7   21.9      2.00   54.0   
623181  412    서대문    0   0.0    0     0  0.0  30.6  107.3      1.19  356.5   
623182  411     마포    0   0.0    0     0  0.0  29.4  319.9      0.39  315.2   
623183  405   양천 *    0   0.0    0     0  0.0  32.2  344.1      1.50  352.4   
623184  404     강서    0   0.0    0     0  0.0  30.4  311.5      1.61  349.8   
623185  110  김포(공)    .   0.0    .     .  0.0  30.5  270.0      1.00   40.0   
623186  423     구로    0   0.0    0     0  0.0    31  321.7 

#### 풍속 라벨링
- 바람조금(0.0~1.5) 0
- 바람적당히(1.6~7.9) 1
- 바람많이(8.0~13.8) 2

In [307]:
weather3['풍속라벨링'] = weather3['풍속60(m/s)'].apply(lambda x: 0 if (0.0 <= x and x <= 1.5)
                                                else 1 if (1.5 < x and x < 8.0)
                                                else 2)

In [309]:
weather3.describe()

ID           강수60            일강수            풍향1  \
count  623198.000000  623198.000000  623198.000000  623198.000000   
mean      413.847224       0.287312       3.345996     154.956247   
std       122.581636       1.542788       9.666424     106.400965   
min       108.000000       0.000000       0.000000       0.000000   
25%       405.000000       0.000000       0.000000      66.000000   
50%       412.000000       0.000000       0.000000     140.400000   
75%       419.000000       0.000000       0.000000     239.100000   
max       889.000000      48.500000      72.500000     360.000000   

            풍속1(m/s)           풍향10      풍속10(m/s)          우천라벨링  \
count  623198.000000  623198.000000  623198.000000  623198.000000   
mean        1.517316     157.738625       1.522099       0.107871   
std         1.248253     104.529735       1.123327       0.357565   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.610000      70.800000       0.690000       0.000000   
50%         1.310000     142.800000       1.310000       0.000000   
75%         2.110000     238.900000       2.110000       0.000000   
max        14.110000     360.000000      12.610000       2.000000   

           풍속60(m/s)          풍속라벨링  
count  623198.000000  623198.000000  
mean        1.517316       0.462221  
std         0.735760       0.498571  
min         0.395425       0.000000  
25%         0.937425       0.000000  
50%         1.387919       0.000000  
75%         1.925564       1.000000  
max         4.206794       1.000000

-------------------

In [310]:
weather2.to_csv('weather2.csv')
weather3.to_csv('weather3.csv')